sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -132,6 +135,61 @@ private[sql] object ArrowConverters {
     }
   }
 
+  private[sql] def fromPayloadIterator(iter: Iterator[ArrowPayload]): Iterator[InternalRow] = {
+    new Iterator[InternalRow] {
+      private val _allocator = new RootAllocator(Long.MaxValue)
+      private var _reader: ArrowFileReader = _
+      private var _root: VectorSchemaRoot = _
+      private var _index = 0
+
+      loadNextBatch()
+
+      override def hasNext: Boolean = _root != null && _index < _root.getRowCount
+
+      override def next(): InternalRow = {
+        val fields = _root.getFieldVectors.asScala
+
+        val genericRowData = fields.map { field =>
+          field.getAccessor.getObject(_index)
+        }.toArray[Any]
```

Comment:
How about using `SpecificInternalRow`? I think that it could eliminate some boxing/unboxing. The following is a snippet for this usage.

```java
        val fieldTypes = fields.map { field =>
          field match {
            case NullableIntVector => IntegerType
            case NullableFloat8Vector => DoubleType
            ...
          }
        }
        val row = new SpecificInternalRow(fieldTypes)
        fields.zipWithIndex.map { case (field, i) =>
          field match {
            case NullableIntVector =>
              row.setInt(i, field.asInstanceOf[NullableIntVector].getAccessor.get(_index)) 
            case NullableFloat8Vector => LongType
              row.setDouble(i, field.asInstanceOf[NullableFloat8Vector].getAccessor.get(_index))
            ...  
          }
        }
```